src/CoreWCF.Primitives/src/CoreWCF/Security/SamlSerializer.cs

```diff
@@ -1,9 +1,115 @@
+using Microsoft.IdentityModel.Tokens.Saml;
+
 namespace CoreWCF.Security
 {
     public class SamlSerializer
     {
+        private readonly MSSAmlSerializer _mSSamlSerializer;
+        public SamlSerializer()
+        {
+            _mSSamlSerializer = new MSSAmlSerializer();
+        }
+
+        internal IdentityModel.Tokens.SamlSecurityToken ReadToken(XmlDictionaryReader reader, SecurityTokenSerializer keyInfoSerializer, SecurityTokenResolver outOfBandTokenResolver)
+        {
+            if (reader == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(reader));
+            }
+
+            var assertion = _mSSamlSerializer.ReadAssertion(reader);
+            if (assertion == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLUnableToLoadAssertion)));
+            }
+
+            SecurityKeyIdentifier securityKeyIdentifier = CoreWCF.Security.SecurityUtils.CreateSecurityKeyIdentifier(assertion.Signature.KeyInfo);
+            SecurityKey verificationKey = SamlSerializer.ResolveSecurityKey(securityKeyIdentifier, outOfBandTokenResolver); ;
+            if (verificationKey == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLUnableToResolveSignatureKey, "issuer")));
+            }
+
+            var samlSecurityToken = new IdentityModel.Tokens.SamlSecurityToken(assertion)
+            {
+                SigningToken = SamlSerializer.ResolveSecurityToken(securityKeyIdentifier, outOfBandTokenResolver)
+            };
+            if (samlSecurityToken.SigningToken == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlSigningTokenNotFound)));
+            }
+
+            foreach (SamlStatement stmt in assertion.Statements)
+            {
+                SamlSubjectStatement samlSubject = (SamlSubjectStatement)stmt;
+                SecurityKeyIdentifier keyIdentifier = null;
+                SecurityToken securityToken = null;
+                if (samlSubject.Subject.KeyInfo != null)
+                {
+                    keyIdentifier = CoreWCF.Security.SecurityUtils.CreateSecurityKeyIdentifier(samlSubject.Subject.KeyInfo);
+                    securityToken = SamlSerializer.ResolveSecurityToken(securityKeyIdentifier, outOfBandTokenResolver);
+                }
+                InternalSamlSubjectStatement internalSamlSubjectStatement = new InternalSamlSubjectStatement(samlSubject, keyIdentifier, securityToken);
+                samlSecurityToken.SamlStatements.Add(internalSamlSubjectStatement);
+            }
+
+            //do a final validation on sign.
+            assertion.Signature.SignedInfo.Verify(Microsoft.IdentityModel.Tokens.CryptoProviderFactory.Default);
```

Comment:
I don't think the original code actually verified the signature. There's no real harm doing it here other than consuming CPU.